File: Week3/684.py
"""
https://leetcode-cn.com/problems/redundant-connection/
并查集解法
"""
class Solution:
    def findRedundantConnection(self, edges):
        self.lst={i:i for i in range(1,len(edges)+1)}
        self.size={i:0 for i in range(1,len(edges)+1)}
        for u,v in edges:
            con=self.connect(u,v)
            if not con:
                if u>v:
                    return [v,u]
                else:
                    return [u,v]
    def getparent(self,node):
        if node!=self.lst[node]:
            return self.getparent(self.lst[node])
        return node
    def connect(self,u,v):
        uparent=self.getparent(u)
        vparent=self.getparent(v)
        if uparent==vparent:
            return False
        usize=self.size[uparent]
        vsize=self.size[vparent]
        if usize>=vsize:
            self.lst[vparent]=uparent
            if usize==vsize:
                self.size[uparent]+=1
        else:
            self.lst[uparent]=vparent
        return True

File: Week7/684.java
class Solution {
    public int[] findRedundantConnection(int[][] edges) {
        int n=edges.length;
        int[] parent=new int[n+1];
        for (int i=1;i<=n;i++){
            parent[i]=i;
        }
        for (int i=0;i<n;i++){
            boolean con=connect(parent,edges[i][0],edges[i][1]);
            if (!con){
                if(edges[i][0]>edges[i][1]){
                    return new int[]{edges[i][1],edges[i][0]};
                }
                return new int[]{edges[i][0],edges[i][1]};
            }
        }
        return new int[]{};
    }
    public boolean connect(int[] parent,int u,int v){
        int rootu=getparent(parent,u);
        int rootv=getparent(parent,v);
        if (rootu!=rootv){
            parent[rootv]=rootu;
            return true;
        }
        return false;
    }
    private int getparent(int[] parent,int node){
        if (node!=parent[node]){
            return getparent(parent,parent[node]);
        }
        return node;
    }
}